win32: add an option to control window title bar state

kasper93 opened the pull request.

natural-harmonia-gropius commented:
There still a bar.
![image](https://user-images.githubusercontent.com/50797982/226174610-e9b4a2e3-2b77-449c-bb16-efa401c347ff.png)
And I use `set title-bar no` in console, it doesn't take effect. After `cycle border` twice, it works.

kasper93 commented:
> There still a bar.

Yes, there is a border. DMW makes left, right and bottom border invisible, but it is there, just it is visible on top.

This is just how Windows works, but we can reduce top window border size to imitate DWM invisible borders. I've added a commit. One side effect is that border grab area is different on top, because simply the border itself is smaller, compared to others.

I think for app like mpv it make sense to reduce top border size, so I made the change.

> And I use set title-bar no in console, it doesn't take effect. After cycle border twice, it works.

Good catch. I missed on dynamic settings update check. Fixed now.

kasper93 commented:
Rebased and as a bonus added `--window-corners` option. This PR is now ready, other improvements might follow in next ones.

kasper93 commented:
Windows builds are available here, if anyone wants to try https://github.com/mpv-player/mpv/actions/runs/5932431497?pr=11469

natural-harmonia-gropius commented:
Border looks pretty good now 👍. But no transition animation when `--title-bar=no`

kasper93 commented:
> But no transition animation when --title-bar=no

What animation?

natural-harmonia-gropius commented:
https://github.com/mpv-player/mpv/assets/50797982/056b1892-88ca-435d-88dc-3ad267d67c3f

https://github.com/melak47/BorderlessWindow

This demo works as I expect, using `WS_POPUP` with `WS_CAPTION`.

kasper93 commented:
Nah, it completely disables non-client area drawing by DWM, with a shadow restore. But there is no window outline and the shadow is not visible in fact.

I guess I can enable `WS_CAPTION` and just exclude whole title bar, not just border.

kasper93 commented:
@natural-harmonia-gropius Try this version, should be fine now. As a bonus we don't have to ask DWM. Hopefully, our geometry is correct ;p

kasper93 commented:
Thanks for test. I guess it is ready for review and merge now.

MG240 commented:
I'm eagerly waiting for this to be merged.

Dudemanguy commented on DOCS/man/options.rst:
```diff
@@ -3135,6 +3135,11 @@ Window
     Play video with window border and decorations. Since this is on by
     default, use ``--no-border`` to disable the standard window decorations.
 
+``--title-bar``, ``--no-title-bar``
+    Play video with window title bar. Since this is on by default,
+    use ``--no-title-bar`` to hide title bar. ``--no-border`` option takes
+    precedence. (Windows only)
```
```suggestion
``--title-bar``, ``--no-title-bar``
    (Windows only)
    Play video with window title bar. Since this is on by default,
    use ``--no-title-bar`` to hide title bar. ``--no-border`` option takes
    precedence.
```
Seems to be the convention (well mostly).

kasper93 replied:
In this section of manual it seems to be at the first line. I would change it in separate commit. Look at those `(macOS only)` and `(Windows only)`

Dudemanguy replied:
What do you mean? Since you're adding this, I was just suggesting that the `(Windows only)` should be on the first line like the rest.

kasper93 replied:
ah, I'm tired. I read it other way around.

kasper93 replied:
fixed.

Dudemanguy commented on video/out/w32_common.c:
```diff
@@ -822,6 +840,10 @@ static void update_window_style(struct vo_w32_state *w32)
+
+    w32->top_border_reduce = (w32->opts->border && !w32->opts->title_bar &&
+                              !w32->current_fs)
+                                ? get_invisible_border_size(w32) : 0;
```
```suggestion
   w32->top_border_reduce = (w32->opts->border && !w32->opts->title_bar &&
                              !w32->current_fs) ? get_invisible_border_size(w32) : 0;
```
I dunno. Looks better imo. 85 characters isn't that long.

kasper93 replied:
changed
